src/gluonnlp/lr_scheduler/reduce_lr_on_plateau.py

```diff
@@ -0,0 +1,210 @@
+        self.is_better = None
+        self.eps = eps
+        self.last_epoch = -1
+        self._init_is_better(mode=mode,
+                             threshold=threshold,
+                             threshold_mode=threshold_mode)
+        self._reset()
+
+    def _reset(self):
+        r"""Resets num_bad_epochs counter and cooldown counter."""
+        self.best = self.mode_worse
+        self.cooldown_counter = 0
+        self.num_bad_epochs = 0
+
+    def step(self, metric, epoch=None):
+        r"""Function to be executed after model evaluation
+
+        Parameters
+        ----------
+        metric : float
+            Current metric value to mesure model performance.
+        epoch : int, default None
+            Current epoch. If None, it is managed internally.
+
+        """
+        current = float(metric)
+        if epoch is None:
+            epoch = self.last_epoch = self.last_epoch + 1
+        self.last_epoch = epoch
+
+        if self.is_better(current, self.best):
+            self.best = current
+            self.num_bad_epochs = 0
+        else:
+            self.num_bad_epochs += 1
+
+        if self.in_cooldown:
+            self.cooldown_counter -= 1
+            self.num_bad_epochs = 0
+
+        if self.num_bad_epochs > self.patience:
+            self._reduce_lr(epoch)
+            self.cooldown_counter = self.cooldown
+            self.num_bad_epochs = 0
+
+    def _reduce_lr(self, epoch):
+        old_lr = float(self.trainer.learning_rate)
+        new_lr = max(old_lr * self.factor, self.min_lr)
+        if old_lr - new_lr > self.eps:
+            self.trainer.set_learning_rate(new_lr)
+            if self.verbose:
+                print('Epoch {:5d}: reducing learning rate'
+                      ' {} to {}.'.format(epoch, old_lr, new_lr))
+
+    @property
+    def in_cooldown(self):
+        return self.cooldown_counter > 0
+
+    def _cmp(self, mode, threshold_mode, threshold, a, best):
+        if mode == 'min' and threshold_mode == 'rel':
+            rel_epsilon = 1. - threshold
+            return a < best * rel_epsilon
+
+        elif mode == 'min' and threshold_mode == 'abs':
+            return a < best - threshold
+
+        elif mode == 'max' and threshold_mode == 'rel':
+            rel_epsilon = threshold + 1.
+            return a > best * rel_epsilon
+
+        else:  # mode == 'max' and epsilon_mode == 'abs':
+            return a > best + threshold
+
+    def _init_is_better(self, mode, threshold, threshold_mode):
+        if mode not in {'min', 'max'}:
+            raise ValueError('mode ' + mode + ' is unknown!')
+        if threshold_mode not in {'rel', 'abs'}:
+            raise ValueError('threshold mode ' + threshold_mode +
+                             ' is unknown!')
+
+        if mode == 'min':
+            self.mode_worse = np.Inf
+        else:  # mode == 'max':
+            self.mode_worse = -np.Inf
+
+        self.is_better = partial(self._cmp, mode, threshold_mode, threshold)
```

Comment:
Why do you want to use partial? `mode, threshold_mode, threshold` are simply class variables, so you can simply use them by adding `self.` to the prefix.

Reply:
Right.  I will fix.